src/services/RPCSelector.ts

```diff
@@ -0,0 +1,69 @@
+import { JsonRpcProvider, Block } from '@ethersproject/providers';
+import { isTimestampMoreRecentThan } from '../utils/helpers';
+
+interface ProviderComparand {
+  blockNumber: number;
+  url: string;
+}
+
+class RPCSelector {
+  readonly urls: string[];
+  readonly preferredProviderUrl: string;
+  readonly rpcRequestTimeout: number;
+  readonly maxBlockAge: number;
+
+  constructor(urls: string | string[], rpcRequestTimeout = 15000, maxBlockAge = 60) {
+    this.urls = typeof urls === 'string' ? urls.split(',') : urls;
+    this.preferredProviderUrl = this.urls[0];
+    this.rpcRequestTimeout = rpcRequestTimeout;
+    this.maxBlockAge = maxBlockAge;
+  }
+
+  async apply(): Promise<string> {
+    if (this.urls.length === 1 || (await this.isPreferredProviderUpToDate())) {
```

Comment:
Can we have a simple for loop here?

```
async apply(): Promise<string> {
	if (this.urls.length < 2) {
		// return the only URL
		return this.urls[0];
	}

	for (let i = 0; i < this.urls.length; ++i) {
		if (await this.isPreferredProviderUpToDate(this.urls[i])) {
			return this.urls[i];
		}
	}
}
```

Reply:
No. At least, not at the moment.
We're running this at runtime instead of a worker, so It needs to be fast. Some RPCs take about 30 seconds or so to respond.
Even with the timeout configured to 500 ms, it's still a problem if there are multiple RPCs because it will take roughly 1 second to run.
You even proposed that we check the block numbers from the rest of the RPCs instead of the timestamp.
The solution you're proposing is unfeasible right now.